contrib/hadoop-store-builder/src/java/voldemort/store/readonly/mr/azkaban/GobblinDistcpJob.java

```diff
@@ -0,0 +1,274 @@
+        info(path + " is scheduled to be deleted on exit.");
+    }
+
+    private void checkPermission(FileSystem fs, Path path, boolean checkWritePermission) throws Exception {
+        FsPermission perm = fs.getFileStatus(path).getPermission();
+        FsAction u = perm.getUserAction();
+        FsAction g = perm.getGroupAction();
+        FsAction o = perm.getOtherAction();
+        boolean changed = false;
+
+        // Check read permission
+        if (props.getBoolean(VoldemortBuildAndPushJob.PUSH_CDN_READ_BY_GROUP, false)) {
+            if (!g.implies(FsAction.READ_EXECUTE)) {
+                g = g.or(FsAction.READ_EXECUTE);
+                changed = true;
+            }
+        }
+
+        if (props.getBoolean(VoldemortBuildAndPushJob.PUSH_CDN_READ_BY_OTHER, false)) {
+            if (!o.implies(FsAction.READ_EXECUTE)) {
+                o = o.or(FsAction.READ_EXECUTE);
+                changed = true;
+            }
+        }
+
+        // Check write permission
+        if (checkWritePermission) {
+            if (props.getBoolean(VoldemortBuildAndPushJob.PUSH_CDN_WRITTEN_BY_GROUP, false)) {
+                if (!g.implies(FsAction.WRITE)) {
+                    g = g.or(FsAction.WRITE);
+                    changed = true;
+                }
+            }
+
+            if (props.getBoolean(VoldemortBuildAndPushJob.PUSH_CDN_WRITTEN_BY_OTHER, false)) {
+                if (!o.implies(FsAction.WRITE)) {
+                    o = o.or(FsAction.WRITE);
+                    changed = true;
+                }
+            }
+        }
+
+        if (changed) {
+            FsPermission desiredPerm = new FsPermission(u, g, o);
+            fs.setPermission(path, desiredPerm);
+            if (!fs.getFileStatus(path).getPermission().equals(desiredPerm)) {
+                throw new RuntimeException("Failed to set permission for " + path + " from: " + perm + " to: " + desiredPerm);
+            }
+            info("for path " + path + ", permissions changed from: " + perm + " to: " + desiredPerm);
+        }
+    }
+
+    private void addPermissionsToParentDirs(FileSystem fs, String target, String pathPrefix) throws Exception {
+        String pathFromRoot = removeLeadingSlash(extractPathFromUrl(target));
+        String parent = "";
+
+        for (String seg: pathFromRoot.split("/")) {
+            parent = parent + "/" + seg;
+            Path parentPath = new Path(parent);
+
+            if (parent.length() <= pathPrefix.length()) {
+                checkPermission(fs, parentPath, false);
+            } else {
+                checkPermission(fs, parentPath, true);
+            }
+        }
+    }
+
+    private Map<String, String> buildCdnMap() throws Exception {
+        List<String> pairs = props.getList(VoldemortBuildAndPushJob.PUSH_CDN_CLUSTER);
+        Map<String, String> cdnMap = new HashMap<>(pairs.size());
+
+        for (String pair : pairs) {
+            if (!pair.contains("|")) {
```

Comment:
Is it safe to use '|' and '\\|' as the separators?
As far as I know, '|' is allowed in file name, and filepath could be part of cdn url, right?

Why don't we choose the existing library, such avro or json to avoid this problem?